### PR TITLE
Implement volatile loads and stores

### LINKED ISSUE #166: Volatile loads and stores aren't honored
```rust
#[no_mangle]
pub unsafe fn assign_answer(x: *mut u32) {
    x.read_volatile();
    x.write_volatile(42);
    x.write_volatile(42);
}
```

In the code generated by cg_llvm, all volatile loads and stores are preserved.
```
0000000000001100 <assign_answer>:
    1100:       8b 07                   mov    (%rdi),%eax
    1102:       c7 07 2a 00 00 00       movl   $0x2a,(%rdi)
    1108:       c7 07 2a 00 00 00       movl   $0x2a,(%rdi)
    110e:       c3                      ret    
    110f:       90                      nop
```

In the code generated by cg_gcc, the load whose result is discarded is eliminated, and the overlapping stores are combined into one.
```
0000000000001100 <assign_answer>:
    1100:       c7 07 2a 00 00 00       movl   $0x2a,(%rdi)
    1106:       c3                      ret    
    1107:       66 0f 1f 84 00 00 00    nopw   0x0(%rax,%rax,1)
    110e:       00 00 
```

### PR DESCRIPTION
**Pull request by yvt**
Fixes #166.

### CODE REVIEW

File: src/builder.rs [hunk 1]
```diff
@@ -654,24 +654,12 @@ impl<'a, 'gcc, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'gcc, 'tcx> {
...
-        // the current basic block. Otherwise, it could be used in another basic block, causing a
-        // dereference after a drop, for instance.
-        // TODO(antoyo): handle align of the load instruction.
-        let deref = ptr.dereference(None).to_rvalue();
-        let value_type = deref.get_type();
-        unsafe { RETURN_VALUE_COUNT += 1 };
-        let loaded_value = function.new_local(None, value_type, &format!("loadedValue{}", unsafe { RETURN_VALUE_COUNT }));
-        block.add_assignment(None, loaded_value, deref);
-        loaded_value.to_rvalue()
+        // TODO: handle non-natural alignments
```
Comment (antoyo) [inline src/builder.rs:657]: The rust repo requires that TODO should be annotated with a name. If you don't plan to fix this later, just put my username there.

File: src/builder.rs [hunk 2]
```diff
@@ -799,17 +787,32 @@ impl<'a, 'gcc, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'gcc, 'tcx> {
...
-        self.llbb().add_assignment(None, aligned_destination, val);
-        // TODO(antoyo): handle align and flags.
+
+        let align = if flags.contains(MemFlags::UNALIGNED) { 1 } else { align.bytes() };
+        let mut modified_destination_type = destination_type.get_aligned(align);
+        if flags.contains(MemFlags::VOLATILE) {
+            modified_destination_type = modified_destination_type.make_volatile();
+        }
+
+        // FIXME: The type checking in `add_assignment` removes only one
```
Comment (antoyo) [inline src/builder.rs:804]: Same for `FIXME`.

File: src/builder.rs [hunk 3]
```diff
@@ -799,17 +787,32 @@ impl<'a, 'gcc, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'gcc, 'tcx> {
...
+        // `int → int __attribute__((aligned(1)))` are considered legal but
+        // `int → volatile int __attribute__((aligned(1)))` is not. This seems
+        // like a bug in libgccjit. The easiest way to work around this is to
+        // bitcast `val` to have the matching qualifiers.
+        val = self.cx.context.new_bitcast(None, val, modified_destination_type);
+
+        let modified_ptr = self.cx.context.new_bitcast(None, ptr, modified_destination_type.make_pointer());
+        let modified_destination = modified_ptr.dereference(None);
+        self.llbb().add_assignment(None, modified_destination, val);
+        // TODO: handle `MemFlags::NONTEMPORAL`.
```
Comment (antoyo) [inline src/builder.rs:815]: Same here.

File: src/builder.rs [hunk 4]
```diff
@@ -1296,6 +1299,33 @@ impl<'a, 'gcc, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'gcc, 'tcx> {
...
+        let function = block.get_function();
+
+        // NOTE: instead of returning the dereference here, we have to assign it to a variable in
+        // the current basic block. Otherwise, it could be used in another basic block, causing a
+        // dereference after a drop, for instance.
+        let deref = ptr.dereference(None).to_rvalue();
+        let value_type = deref.get_type();
+
+        // If `is_volatile == true`, convert `ptr` to `volatile value_type *`
+        // and try dereference again. (libgccjit doesn't provide a way to get
```
Comment (antoyo) [inline src/builder.rs:1313]: I'm not sure if that's what you're talking about, but an alternative is to do:
```rust
ptr.get_type().get_pointee()
```

**Review by antoyo — COMMENTED** [review]
LGTM
Thanks!

**Comment by antoyo** [comment]
The CI failed for libgccjit12, btw. Not sure if it's normal.

**Comment by yvt** [comment]
Not normal, I think. Also, I found that volatile reads are still being eliminated by `dse` pass, which I'm looking into.

```rust
p_count.write_volatile(0);
STORAGE.add(0).read_volatile();
STORAGE.add(PAGE_SIZE).read_volatile();
STORAGE.add(0).write_volatile(1);
```

`fake.c.040t.mergephi1`:

```
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_14write_volatile17hb4eb8141481df2c1E
  _ZN8volatile5COUNT17h55d9a409972f8391E = 0;
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_3add17h994bc78b46d6ca3bE
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_13read_volatile17hb2bc363effab5c7cE
  loadedValue3_87 = MEM[(volatile unsigned char *)ptrReturnValue11_69];
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_3add17h994bc78b46d6ca3bE
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_13read_volatile17hb2bc363effab5c7cE
  loadedValue3_97 = MEM[(volatile unsigned char *)ptrReturnValue11_69 + 32768B];
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_3add17h994bc78b46d6ca3bE
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_14write_volatile17h2a645162d347c6a8E
  MEM[(volatile unsigned char *)ptrReturnValue11_69] = 1;
```

`fake.c.041t.dse1`:

```
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_14write_volatile17hb4eb8141481df2c1E
  _ZN8volatile5COUNT17h55d9a409972f8391E = 0;
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_3add17h994bc78b46d6ca3bE
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_13read_volatile17hb2bc363effab5c7cE
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_3add17h994bc78b46d6ca3bE
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_13read_volatile17hb2bc363effab5c7cE
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_3add17h994bc78b46d6ca3bE
  # DEBUG INLINE_ENTRY _ZN4core3ptr7mut_ptr31__LT_impl_u20__BP_mut_u20_T_GT_14write_volatile17h2a645162d347c6a8E
  MEM[(volatile unsigned char *)ptrReturnValue11_69] = 1;
```

File: src/builder.rs (hunk 4 above)
Comment (yvt) [reply src/builder.rs:1313]: Thanks, that's exactly what I was looking for!

**Comment by kellda** [comment]
I just got bit by this bug. Is there any plan to fix it ?

**Comment by antoyo** [comment]
@kellda: I will try to take a look in the upcomming weeks.

**Comment by antoyo** [comment]
@kellda: This was implemented in #572. Please tell me if that works for you.

**Comment by kellda** [comment]
Yes it works, thank you

**Pull request closed without merging**